### PR TITLE
Implement WINFF.FT.S.HBT.2, 3, 4

### PR DESCRIPTION
**Pull request by krlinus**
heartbeat cases

### CODE REVIEW

File: src/harness/testcases/heartbeat_testcase.py [hunk 1]
```diff
@@ -78,6 +78,198 @@ def test_10_9_4_1_1_1(self):
...
+    self.assertEqual(len(response), 3)
+    for response_num, resp in enumerate(response):
+      self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+      self.assertEqual(resp['grantId'], grant_ids[response_num])
+      transmit_expire_time = datetime.strptime(resp['transmitExpireTime'],
+                                               '%Y-%m-%dT%H:%M:%SZ')
+      self.assertLess(datetime.utcnow(), transmit_expire_time)
+      self.assertLessEqual((transmit_expire_time -
+                       datetime.utcnow()).total_seconds(), 240)
+      self.assertLess(transmit_expire_time, grant_expire_time[response_num])
```
Comment (janetliuboston) [inline src/harness/testcases/heartbeat_testcase.py]: The transmit expire time and grant expire time times could be the same: "transmitExpireTime shall be not later than grantExpireTime".

assertLessEqual(transmit_expire_time, grant_expire_time[response_num])

File: src/harness/testcases/heartbeat_testcase.py [hunk 2]
```diff
@@ -78,6 +78,198 @@ def test_10_9_4_1_1_1(self):
...
+    response = self._sas.Heartbeat(request)['heartbeatResponse'][0]
+    # Check the heartbeat response
+    self.assertEqual(response['cbsdId'], cbsd_id)
+    self.assertEqual(response['grantId'], grant_id)
+    transmit_expire_time = datetime.strptime(response['transmitExpireTime'],
+                                             '%Y-%m-%dT%H:%M:%SZ')
+    self.assertLess(datetime.utcnow(), transmit_expire_time)
+    self.assertLessEqual((transmit_expire_time -
+                     datetime.utcnow()).total_seconds(), 240)
+    self.assertLess(transmit_expire_time, grant_expire_time)
```
Comment (janetliuboston) [inline src/harness/testcases/heartbeat_testcase.py]: Same comment as above, transmit_expire_time and grant_expire_time could be the same:
 self.assertLessEqual(transmit_expire_time, grant_expire_time)

File: src/harness/testcases/heartbeat_testcase.py [hunk 3]
```diff
@@ -78,6 +78,198 @@ def test_10_9_4_1_1_1(self):
...
+        open(os.path.join('testcases', 'testdata', 'device_b.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    device_c = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_c.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    request = {'registrationRequest': [device_a, device_b, device_c]}
+    response = self._sas.Registration(request)['registrationResponse']
+    cbsd_ids = []
+    for resp in response:
+      self.assertEqual(resp['response']['responseCode'], 0)
```
Comment (janetliuboston) [inline src/harness/testcases/heartbeat_testcase.py]: Perhaps use 4 spaces for tab?

File: src/harness/testcases/heartbeat_testcase.py [hunk 4]
```diff
@@ -78,6 +78,198 @@ def test_10_9_4_1_1_1(self):
...
+    # Heartbeat requests with grantRenew set to True
+    heartbeat_request = []
+    for cbsd_id, grant_id in zip(cbsd_ids, grant_ids):
+      heartbeat_request.append({
+          'cbsdId': cbsd_id,
+          'grantId': grant_id,
+          'operationState': 'GRANTED',
+          'grantRenew': True
+      })
+    request = { 'heartbeatRequest': heartbeat_request}
```
Comment (janetliuboston) [inline src/harness/testcases/heartbeat_testcase.py]: Remove extra space after curly bracket?
 { 'heartbeatRequest' ->  {'heartbeatRequest'

File: src/harness/testcases/heartbeat_testcase.py [hunk 5]
```diff
@@ -78,6 +78,198 @@ def test_10_9_4_1_1_1(self):
...
+    # Check the heartbeat response
+    for response_num, resp in enumerate(response):
+      transmit_expire_time = datetime.strptime(resp['transmitExpireTime'],
+                                               '%Y-%m-%dT%H:%M:%SZ')
+      self.assertEqual(resp['cbsdId'], cbsd_ids[response_num])
+      self.assertEqual(resp['grantId'], grant_ids[response_num])
+      self.assertLess(datetime.utcnow(), transmit_expire_time)
+      self.assertLessEqual((transmit_expire_time -
+                       datetime.utcnow()).total_seconds(), 240)
+      self.assertLess(transmit_expire_time, grant_expire_times[response_num])
```
Comment (janetliuboston) [inline src/harness/testcases/heartbeat_testcase.py]: The times could be the same.
self.assertLessEqual(transmit_expire_time, grant_expire_times[response_num])

File: src/harness/testcases/heartbeat_testcase.py (hunk 3 above)
Comment (krlinus) [inline src/harness/testcases/heartbeat_testcase.py]: Our formatting tool always uses 2 spaces for indenting into functions and loops. However, other issues should be ironed out now

**Pull request merged**